packages/beacon-node/src/util/clock.ts

```diff
@@ -97,51 +115,63 @@ export class Clock extends EventEmitter implements IClock {
     const currentSlot = this.currentSlot;
     const nextSlotTime = computeTimeAtSlot(this.config, currentSlot + 1, this.genesisTime) * 1000;
-    return nextSlotTime - Date.now() < MAXIMUM_GOSSIP_CLOCK_DISPARITY ? currentSlot + 1 : currentSlot;
+    return nextSlotTime - Date.now() < MAXIMUM_GOSSIP_CLOCK_DISPARITY_MS ? currentSlot + 1 : currentSlot;
   }
 
   get currentEpoch(): Epoch {
     return computeEpochAtSlot(this.currentSlot);
   }
 
-  /** Returns the slot if the internal clock were advanced by `toleranceSec`. */
-  slotWithFutureTolerance(toleranceSec: number): Slot {
-    // this is the same to getting slot at now + toleranceSec
-    return getCurrentSlot(this.config, this.genesisTime - toleranceSec);
+  /** Returns the slot if the internal clock were advanced by `toleranceMs`. */
+  slotWithFutureTolerance(toleranceMs: number): Slot {
+    const currentSlot = this.currentSlot;
+    const currentSlotStartTime = computeTimeAtSlot(this.config, currentSlot, this.genesisTime) * 1000;
+    const now = Date.now();
+    return toleranceMs > 0 && now > currentSlotStartTime && now - currentSlotStartTime <= toleranceMs
+      ? this.currentSlot - 1
+      : this.currentSlot;
   }
 
-  /** Returns the slot if the internal clock were reversed by `toleranceSec`. */
-  slotWithPastTolerance(toleranceSec: number): Slot {
-    // this is the same to getting slot at now - toleranceSec
-    return getCurrentSlot(this.config, this.genesisTime + toleranceSec);
+  /** Returns the slot if the internal clock were reversed by `toleranceMs`. */
+  slotWithPastTolerance(toleranceMs: number): Slot {
+    const currentSlot = this.currentSlot;
+    const currentSlotEndTime = computeTimeAtSlot(this.config, currentSlot + 1, this.genesisTime) * 1000 - 1;
+    const now = Date.now();
+    return toleranceMs > 0 && now < currentSlotEndTime && currentSlotEndTime - now < toleranceMs
+      ? currentSlot + 1
+      : currentSlot;
   }
 
-  /**
-   * Check if a slot is current slot given MAXIMUM_GOSSIP_CLOCK_DISPARITY.
-   */
-  isCurrentSlotGivenGossipDisparity(slot: Slot): boolean {
+  isCurrentSlotGivenTolerance(slot: Slot, pastToleranceMs: number, futureToleranceMs?: number): boolean {
     const currentSlot = this.currentSlot;
     if (currentSlot === slot) {
       return true;
     }
-    const nextSlotTime = computeTimeAtSlot(this.config, currentSlot + 1, this.genesisTime) * 1000;
-    // we're too close to next slot, accept next slot
-    if (nextSlotTime - Date.now() < MAXIMUM_GOSSIP_CLOCK_DISPARITY) {
-      return slot === currentSlot + 1;
-    }
-    const currentSlotTime = computeTimeAtSlot(this.config, currentSlot, this.genesisTime) * 1000;
-    // we've just passed the current slot, accept previous slot
-    if (Date.now() - currentSlotTime < MAXIMUM_GOSSIP_CLOCK_DISPARITY) {
-      return slot === currentSlot - 1;
-    }
+    pastToleranceMs = pastToleranceMs ?? 0;
+    futureToleranceMs = futureToleranceMs ?? pastToleranceMs ?? 0;
```

Comment:
the ?? 0 isn't needed

Can also use [`??=`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment) like:

```suggestion
    futureToleranceMs ??= pastToleranceMs;
```